### ISSUE TITLE
UVDoc模型自适应宽高咨询

### ISSUE THREAD

**Issue by sxj731533730** [Tencent/ncnn#6345]
## detail | 详细描述 | 詳細な説明

https://sxj731533730.obs.cn-east-317.qdrgznjszx.com/code.zip

群主大佬 我转了paddleocr中的UVDoc.onnx模型 
"pnnx UVDoc.onnx inputshape=[1,3,320,320] inputshape2=[1,3,640,640]"，同时修改了UVDoc.ncnn.param为 sxjUVDoc.ncnn.param可以进行"main.cpp" 自动矫正图片了

但是在 Interp层 我仍然需要根据输入图片的hw 去手动设置3= 4=两个参数，我希望他能根据我的input输入宽度 自动获取hw 而避免我手动设置它;  

其它文件都是原始文件  "UVDoc.pnnx.onnx  UVDoc.ncnn.bin       UVDoc.onnx,UVDoc.pnnx.param,UVDoc.ncnn.param ,UVDoc.pnnx.bin,UVDoc_pnnx.py"

**Comment by nihui** [comment]
https://github.com/Tencent/ncnn/pull/6348

**Comment by nihui** [comment]
- build pnnx from source, with patch (https://github.com/Tencent/ncnn/pull/6348) applied

```shell
pnnx UVDoc.onnx
```

- modify F.interpolate size parameter
- drop noop unsqueeze/squeeze pair around padding

```python
        v_98 = self.convbn2d_42(v_97)
        v_99 = F.relu(input=v_98)
        # v_100 = v_99.unsqueeze(2)
        # v_101 = F.pad(input=v_100, mode='reflect', pad=(2,2,2,2,0,0))
        # v_102 = torch.squeeze(input=v_101, dim=2)
        v_102 = F.pad(input=v_99, mode='reflect', pad=(2,2,2,2))
        v_103 = self.convbn2d_43(v_102)
        v_104 = self.prelu_0(v_103)
        # v_105 = v_104.unsqueeze(2)
        # v_106 = F.pad(input=v_105, mode='reflect', pad=(2,2,2,2,0,0))
        # v_107 = torch.squeeze(input=v_106, dim=2)
        v_107 = F.pad(input=v_104, mode='reflect', pad=(2,2,2,2))
        v_108 = self.conv2d_44(v_107)
        # v_109 = v_0.size(2)
        # v_110 = v_0.size(3)
        # v_111 = torch.cat((v_109, v_110), dim=0)
        # v_112 = [1, 2]
        # v_113 = torch.cat((v_112, v_111), dim=0)
        v_113 = [v_0.size(2), v_0.size(3)]
        v_114 = F.interpolate(input=v_108, size=v_113, align_corners=True, mode='bilinear')
        v_115 = v_114.permute(dims=(0,2,3,1))
        v_116 = F.grid_sample(input=v_0, grid=v_115, align_corners=True, mode='bilinear', padding_mode='zeros')
        return v_116
```

- export new model to pnnx/ncnn

```python
if __name__ == "__main__":
    # print(test_inference())

    net = Model()

    torch.manual_seed(0)
    v_0 = torch.rand(1, 3, 128, 128, dtype=torch.float)
    v_1 = torch.rand(1, 3, 512, 512, dtype=torch.float)

    import pnnx
    pnnx.export(net, "UVDoc_pnnx_fix.pt", (v_0,), (v_1,))
```

- check the inference result

```shell
$ python UVDoc_pnnx_fix_pnnx.py 
/home/nihui/osd/ncnn-nihui-fix/tools/pnnx/build/src/UVDoc_pnnx_fix_pnnx.py:170: UserWarning: `nn.functional.upsample` is deprecated. Use `nn.functional.interpolate` instead.
  v_1 = F.upsample(input=v_0, align_corners=True, mode='bilinear', size=(712,488))
/home/nihui/osd/ncnn-nihui-fix/tools/pnnx/build/src/UVDoc_pnnx_fix_pnnx.py:273: UserWarning: `nn.functional.upsample` is deprecated. Use `nn.functional.interpolate` instead.
  v_104 = F.upsample(input=v_102, size=v_103, align_corners=True, mode='bilinear')
tensor([[[[0.1905, 0.4030, 0.6229,  ..., 0.4021, 0.3439, 0.4884],
          [0.3269, 0.5028, 0.6350,  ..., 0.4006, 0.5047, 0.5137],
          [0.7902, 0.6851, 0.6074,  ..., 0.4399, 0.6744, 0.5100],
          ...,
          [0.2980, 0.4936, 0.5152,  ..., 0.5416, 0.5774, 0.5303],
          [0.2904, 0.4451, 0.3893,  ..., 0.5978, 0.6026, 0.5329],
          [0.2403, 0.3473, 0.2607,  ..., 0.5209, 0.5301, 0.3112]],

         [[0.3784, 0.5647, 0.6466,  ..., 0.6569, 0.4920, 0.4658],
          [0.4239, 0.4410, 0.5653,  ..., 0.6260, 0.3881, 0.3456],
          [0.4789, 0.2891, 0.4495,  ..., 0.5345, 0.2811, 0.2975],
          ...,
          [0.6518, 0.2591, 0.1061,  ..., 0.7958, 0.5617, 0.4211],
          [0.7598, 0.4369, 0.3246,  ..., 0.5911, 0.4291, 0.2759],
          [0.5647, 0.3139, 0.2428,  ..., 0.1292, 0.0952, 0.1258]],

         [[0.3344, 0.3147, 0.4042,  ..., 0.6125, 0.5471, 0.6757],
          [0.3732, 0.3144, 0.3619,  ..., 0.6212, 0.3078, 0.5329],
          [0.3236, 0.2049, 0.2804,  ..., 0.5553, 0.1802, 0.4615],
          ...,
          [0.1641, 0.6079, 0.7458,  ..., 0.9419, 0.6889, 0.3995],
          [0.5629, 0.5995, 0.6029,  ..., 0.7931, 0.5834, 0.3426],
          [0.6227, 0.5136, 0.4754,  ..., 0.6668, 0.4958, 0.3966]]]])

$ python UVDoc_pnnx_fix_ncnn.py 
tensor([[[[0.1903, 0.4456, 0.6675,  ..., 0.4071, 0.3895, 0.4285],
          [0.4279, 0.5540, 0.6466,  ..., 0.3568, 0.4927, 0.5556],
          [0.8553, 0.6907, 0.5871,  ..., 0.3828, 0.6119, 0.6356],
          ...,
          [0.3374, 0.5307, 0.4927,  ..., 0.5361, 0.5651, 0.5247],
          [0.3256, 0.4793, 0.3568,  ..., 0.6140, 0.6064, 0.5398],
          [0.2723, 0.3816, 0.2322,  ..., 0.5956, 0.4971, 0.2732]],

         [[0.4124, 0.5861, 0.6464,  ..., 0.7082, 0.5376, 0.4188],
          [0.4289, 0.4176, 0.5960,  ..., 0.7009, 0.4613, 0.3080],
          [0.4472, 0.2218, 0.5265,  ..., 0.5491, 0.3450, 0.2806],
          ...,
          [0.5725, 0.1810, 0.1108,  ..., 0.7761, 0.5551, 0.4224],
          [0.6841, 0.3606, 0.3156,  ..., 0.6156, 0.4326, 0.2887],
          [0.5253, 0.2873, 0.2649,  ..., 0.0829, 0.0773, 0.1119]],

         [[0.3542, 0.3274, 0.4316,  ..., 0.6564, 0.5395, 0.5106],
          [0.3521, 0.2867, 0.3795,  ..., 0.7412, 0.3895, 0.2993],
          [0.2891, 0.1618, 0.3178,  ..., 0.6029, 0.3527, 0.3373],
          ...,
          [0.2584, 0.6803, 0.7343,  ..., 0.9300, 0.6981, 0.4211],
          [0.5720, 0.6010, 0.6238,  ..., 0.8089, 0.5642, 0.3367],
          [0.6045, 0.5001, 0.4883,  ..., 0.6159, 0.4830, 0.3909]]]])
```

**Comment by nihui** [comment]
disable fp16

```python
    import pnnx
    pnnx.export(net, "UVDoc_pnnx_fix.pt", (v_0,), (v_1,), fp16=False)
```

**Comment by sxj731533730** [comment]
分かりました、恵子さんありがとう
